homeassistant/components/mazda/climate.py

```diff
@@ -0,0 +1,221 @@
+"""Platform for Mazda climate integration."""
+from typing import Any
+
+from pymazda import Client as MazdaAPIClient
+
+from homeassistant.components.climate import ClimateEntity
+from homeassistant.components.climate.const import ClimateEntityFeature, HVACMode
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    ATTR_TEMPERATURE,
+    PRECISION_HALVES,
+    PRECISION_WHOLE,
+    TEMP_CELSIUS,
+    TEMP_FAHRENHEIT,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+from homeassistant.util.temperature import convert as convert_temperature
+
+from . import MazdaEntity
+from .const import DATA_CLIENT, DATA_COORDINATOR, DATA_REGION, DOMAIN
+
+PRESET_DEFROSTER_OFF = "Defroster Off"
+PRESET_DEFROSTER_FRONT = "Front Defroster"
+PRESET_DEFROSTER_REAR = "Rear Defroster"
+PRESET_DEFROSTER_FRONT_AND_REAR = "Front and Rear Defroster"
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the switch platform."""
+    entry_data = hass.data[DOMAIN][config_entry.entry_id]
+    client = entry_data[DATA_CLIENT]
+    coordinator = entry_data[DATA_COORDINATOR]
+    region = entry_data[DATA_REGION]
+
+    async_add_entities(
+        MazdaClimateEntity(client, coordinator, index, region)
+        for index, data in enumerate(coordinator.data)
+        if data["isElectric"]
+    )
+
+
+class MazdaClimateEntity(MazdaEntity, ClimateEntity):
+    """Class for the climate entity."""
+
+    _attr_name = "Climate"
+    _attr_supported_features = (
+        ClimateEntityFeature.TARGET_TEMPERATURE | ClimateEntityFeature.PRESET_MODE
+    )
+    _attr_hvac_modes = [HVACMode.HEAT_COOL, HVACMode.OFF]
+    _attr_preset_modes = [
+        PRESET_DEFROSTER_OFF,
+        PRESET_DEFROSTER_FRONT,
+        PRESET_DEFROSTER_REAR,
+        PRESET_DEFROSTER_FRONT_AND_REAR,
+    ]
+
+    _hvac_mode: HVACMode
+    _target_temperature: float
+    _preset_mode: str
+
+    def __init__(
+        self,
+        client: MazdaAPIClient,
+        coordinator: DataUpdateCoordinator,
+        index: int,
+        region: str,
+    ) -> None:
+        """Initialize Mazda climate entity."""
+        super().__init__(client, coordinator, index)
+
+        self.region = region
+        self._attr_unique_id = self.vin
+
+    async def async_added_to_hass(self) -> None:
+        """When entity is added to hass."""
+        # Whenever a coordinator update happens, update the state of the climate entity.
+        self.async_on_remove(
+            self.coordinator.async_add_listener(self._update_hvac_state)
+        )
+
+        # Perform an initial update of the state.
+        self._update_hvac_state()
```

Comment:
I'm not sure that's needed. Also, since you are inheriting from `CoordinatorEntity`, most of this is taken care of. You just have to implement
```python
@callback
def _handle_coordinator_update(self) -> None:
```
The implementation should be very similar to your `_update_hvac_state ()`, justmake sure to call the `super` implementation at the end (which calls `self.async_write_ha_state()` for you)

Reply:
Thanks, I have implemented your suggestion regarding `_handle_coordinator_update()`. However I think the initial update of the attributes is needed, because without it, I get errors like this when HA tries to read the state:
```
AttributeError: 'MazdaClimateEntity' object has no attribute '_attr_hvac_mode'
```

Reply:
I think you're getting this error because you're not setting `self. _attr_hvac_mode` (and others) in `__init__()`

Reply:
You're right. I updated the code to initialize the state attributes in the constructor, and this allowed me to remove the initial update in `async_added_to_hass()`.